src/primitives/decode/mod.rs

```diff
@@ -0,0 +1,384 @@
+// SPDX-License-Identifier: MIT
+
+//! Decoding of bech32 encoded strings as specified by [BIP-173] (and [BIP-350]).
+//!
+//! A Bech32 string is at most 90 characters long and consists of:
+//!
+//! - The human-readable part, which is intended to convey the type of data, or anything else that
+//!   is relevant to the reader. This part MUST contain 1 to 83 US-ASCII characters.
+//! - The separator, which is always "1".
+//! - The data part, which is at least 6 characters long and only consists of alphanumeric
+//!   characters excluding "1", "b", "i", and "o".
+//!
+//! ## Modules
+//!
+//! > We first describe the general checksummed base32 format called Bech32 and then define
+//! > Segregated Witness addresses using it.
+//!
+//! - mod.rs: Handles parsing the general checksummed base32 format.
+//! - segwit.rs: Handles parsing Segregated Witness addresses.
+//!
+//! [BIP-173]: <https://github.com/bitcoin/bips/blob/master/bip-0173.mediawiki>
+//! [BIP-350]: <https://github.com/bitcoin/bips/blob/master/bip-0350.mediawiki>
+
+pub mod segwit;
+
+use core::{fmt, iter, slice, str};
+
+use crate::primitives::checksum::{self, Checksum};
+use crate::primitives::gf32::Fe32;
+use crate::primitives::hrp::{self, Hrp};
+use crate::primitives::iter::{Fe32IterExt, FesToBytes};
+use crate::write_err;
+
+/// Separator between the hrp and payload (as defined by BIP-173).
+const SEP: char = '1';
+
+/// An HRP string that has been parsed and had the checksum validated.
+///
+/// Pre-parsing an HRP string only checks validity of the characters, it does not validate the
+/// checksum in any way - to validate convert to [`Parsed`].
+#[derive(Debug)]
+pub struct Parsed<'s> {
+    /// The human-readable part, guaranteed to be lowercase ASCII characters.
+    pub(crate) hrp: Hrp,
+    /// This is ASCII byte values of the parsed string, guaranteed to be valid bech32 characters,
+    /// with the checksum removed.
+    pub(crate) data: &'s [u8],
```

Comment:
In 5c9b549a3738cec82f633c14e274854b82edce45:

I'd drop `pub(crate)` from both of these, and add a public accessor to `data`.

Reply:
Woops, we don't even access the data anymore. I left them private for now.